src/Qir/Runtime/lib/Simulators/FullstateSimulator.cpp

```diff
@@ -272,71 +300,87 @@ namespace Quantum
         {
             static TSingleQubitGate op = reinterpret_cast<TSingleQubitGate>(this->GetProc("X"));
             op(this->simulatorId, GetQubitId(q));
+            UnmarkSingleQubit(q);
         }
 
         void ControlledX(long numControls, Qubit controls[], Qubit target) override
         {
             static TSingleQubitControlledGate op = reinterpret_cast<TSingleQubitControlledGate>(this->GetProc("MCX"));
             std::vector<unsigned> ids            = GetQubitIds(numControls, controls);
             op(this->simulatorId, (unsigned)numControls, ids.data(), GetQubitId(target));
+            UnmarkSingleQubit(target);

```

Comment:
The `UnmarkSingleQubit()` is a mechanism common for a number of gates. If we add a few more similar common mechanisms, then a number of gates will contain an identical sequence of calls to those mechanisms.
You may want to consider adding an extra member function containing common mechanisms for multiple gates. And all those gates will be calling that extra member function, and that extra member function will be calling `UnmarkSingleQubit()` and all other common mechanisms.

Reply:
`UnmarkSingleQubit` and `UnmarkQubitList` are the member functions that encapsulate common functionality. Right now there is no other anticipated or planned common functionality to abstract out into a new function, so the new method would be only a call to the unmark functions, which would be redundant abstraction.